bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/GatewayWebTargets.java

```diff
@@ -0,0 +1,523 @@
+        String json = invoke(HttpMethod.GET, shades, null, null);
+        try {
+            return List.of(jsonParser.fromJson(json, Shade[].class));
+        } catch (NullPointerException e) {
+            throw new HubProcessingException("getScenes() missing response");
+        } catch (JsonParseException e) {
+            throw new HubProcessingException("getShades() JsonParseException");
+        }
+    }
+
+    /**
+     * Invoke a call on the hub server to retrieve information or send a command.
+     *
+     * @param method GET or PUT.
+     * @param url the host URL to be called.
+     * @param query the HTTP query parameter.
+     * @param jsonCommand the request command content (as a JSON string).
+     * @return the response content (as a JSON string).
+     * @throws HubProcessingException if something goes wrong.
+     */
+    protected synchronized String invoke(HttpMethod method, String url, @Nullable Query query,
+            @Nullable String jsonCommand) throws HubProcessingException {
+        if (logger.isTraceEnabled()) {
+            if (query != null) {
+                logger.trace("invoke() method:{}, url:{}, query:{}", method, url, query);
+            } else {
+                logger.trace("invoke() method:{}, url:{}", method, url);
+            }
+            if (jsonCommand != null) {
+                logger.trace("invoke() request JSON:{}", jsonCommand);
+            }
+        }
+        Request request = httpClient.newRequest(url).method(method).header("Connection", "close").accept("*/*");
+        if (query != null) {
+            request.param(query.getKey(), query.getValue());
+        }
+        if (jsonCommand != null) {
+            request.header(HttpHeader.CONTENT_TYPE, "application/json").content(new StringContentProvider(jsonCommand));
+        }
+        ContentResponse response;
+        try {
+            response = request.send();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new HubProcessingException(String.format("%s: \"%s\"", e.getClass().getName(), e.getMessage()));
+        } catch (TimeoutException | ExecutionException e) {
+            throw new HubProcessingException(String.format("%s: \"%s\"", e.getClass().getName(), e.getMessage()));
+        }
+        int statusCode = response.getStatus();
+        if (!HTTP_OK_CODES.contains(statusCode)) {
+            logger.warn("invoke() HTTP status:{}, reason:{}", statusCode, response.getReason());
+            throw new HubProcessingException(String.format("HTTP %d error", statusCode));
+        }
+        String jsonResponse = response.getContentAsString();
+        if (logger.isTraceEnabled()) {
+            logger.trace("invoke() response JSON:{}", jsonResponse);
+        }
+        if ((HttpStatus.OK_200 == statusCode) && ((jsonResponse == null) || (jsonResponse.isEmpty()))) {
+            logger.warn("invoke() hub returned no content");
```

Comment:
Any reason to log warning in addition to throwing exception - will exception also be logged?

Reply:
No reason. Exceptions are logged in the handler.

Reply:
It's still there, so redundant logging, and even logged as warning here?

Reply:
Sorry, you removed it, but I saw it also in another line. I have removed it there also.